contracts/interchain-token-service/src/contract/execute/mod.rs

```diff
@@ -150,17 +150,61 @@ fn apply_to_hub(
 
     match message {
         Message::InterchainTransfer(transfer) => {
-            apply_transfer(storage, source_chain, destination_chain, &transfer)
+            apply_to_transfer(storage, source_chain, destination_chain, transfer)
                 .map(Message::InterchainTransfer)?
         }
         Message::DeployInterchainToken(deploy_token) => {
-            apply_token_deployment(storage, &source_chain, &destination_chain, deploy_token)
+            apply_to_token_deployment(storage, &source_chain, &destination_chain, deploy_token)
                 .map(Message::DeployInterchainToken)?
         }
     }
     .then(Result::Ok)
 }
 
+fn apply_to_transfer(
+    storage: &mut dyn Storage,
+    source_chain: ChainNameRaw,
+    destination_chain: ChainNameRaw,
+    mut transfer: InterchainTransfer,
+) -> Result<InterchainTransfer, Error> {
+    interceptors::subtract_supply_amount(storage, &source_chain, &transfer)
+        .and_then(|_| {
+            interceptors::apply_scaling_factor_to_amount(
+                storage,
+                &source_chain,
+                &destination_chain,
+                &mut transfer,
+            )
+        })
+        .and_then(|_| interceptors::add_supply_amount(storage, &destination_chain, &transfer))
+        .map(|_| transfer)
```

Comment:
not that I don't like more functional patterns, but is this necessary here?
Wouldn't it just read
```
interceptors::subtract_supply_amount(...)?;
interceptors::apply_scaling_factor_to_amount(...)?;
interceptors::add_supply_amount(...)?;

Ok(transfer);
```

Reply:
IMO, leaving it functional would encourage people to do one thing at a step in the future when they need to touch this function again. What I want to avoid for this function to happen is people start adding procedural things to it again.

Reply:
would it be possible for these functions to return the appropriate message, i.e. take in `transfer` and pass it back out? That way, the `and_then` wouldn't be wasted and you wouldn't need to map in the end

Reply:
Thought of doing it, but decided to land like this so that it's clearer which interceptors would mutate the messages and which ones would not when someone looks at the apply function.

Reply:
It's a good argument, but is that a concern?

Reply:
No, it's just better readability.

Reply:
I actually disagree, I think the mutability here is a bit hard to follow. I think it would be better if `apply_scaling_factor_to_amount` actually returned the new transfer, instead of mutating. Same with the other functions. We can then save the returned values in this function (if necessary).

Reply:
Returning it hides the fact that it may have changed during the process. Someone looking at the caller's flow can not tell which interceptors transform the message and which do not. It's not an anti-pattern. Rust forces people to use the `mut` keyword for a reason.